java/arcs/sdk/android/storage/ServiceStore.kt

```diff
@@ -103,23 +97,60 @@ class ServiceStore<Data : CrdtData, Op : CrdtOperation, ConsumerData>(
         val service = checkNotNull(storageService)
-        val channel = ParcelableProxyMessageChannel(coroutineContext)
-        service.getLocalData(channel)
-        val flow = channel.asFlow()
-        val modelUpdate =
-            flow.onEach { it.result.complete(true) }
-                .mapNotNull {
-                    it.message.actual as? ProxyMessage.ModelUpdate<Data, Op, ConsumerData>
+        return CompletableDeferred<Data>().also { completable ->
+            service.getLocalData(object : IStorageServiceCallback.Stub() {
+                override fun onProxyMessage(
+                    message: ParcelableProxyMessage,
+                    resultCallback: IResultCallback
+                ) {
+                    scope.launch {
+                        val modelUpdate =
+                            message.actual as? ProxyMessage.ModelUpdate<Data, Op, ConsumerData>
+                        if (modelUpdate != null) {
+                            completable.complete(modelUpdate.model)
+                            resultCallback.onResult(null)
+                        } else {
+                            CrdtException("Wrong message type received").let {
+                                completable.completeExceptionally(it)
+                                resultCallback.onResult(it.toParcelable())
+                            }
+                        }
+                    }
                 }
-                .first()
-        channel.cancel()
-        return modelUpdate.model
+            })
+        }.await()
     }
 
-    override fun on(callback: ProxyCallback<Data, Op, ConsumerData>): Int =
-        proxyCallbacks.register(callback)
+    @Suppress("UNCHECKED_CAST")
+    override fun on(callback: ProxyCallback<Data, Op, ConsumerData>): Int {
+        val service = checkNotNull(storageService)
+        return service.registerCallback(object : IStorageServiceCallback.Stub() {
+            override fun onProxyMessage(
+                message: ParcelableProxyMessage,
+                resultCallback: IResultCallback
+            ) {
+                scope.launch {
+                    val actualMessage = message.actual as? ProxyMessage<Data, Op, ConsumerData>
+                    val result = if (actualMessage != null) {
+                        if (callback.invoke(actualMessage)) null
+                        else CrdtException("Message could not be handled")
+                    } else {
+                        CrdtException("Wrong message type received")
+                    }
+                    resultCallback.onResult(result?.toParcelable())
+                }
+            }
+        })
+    }
 
-    override fun off(callbackToken: Int) = proxyCallbacks.unregister(callbackToken)
+    override fun off(callbackToken: Int) {
+        val service = checkNotNull(storageService)
+        scope.launch {
+            sendQueue.enqueue {
+                service.unregisterCallback(callbackToken)
+            }
+        }
```

Comment:
Ooh, can we add accounting here so that when we reach zero callbacks left to unregister - we can unbind the storage service? (would need to figure out how to re-bind if a callback is added again, though)

Reply:
That is an excellent idea. This PR, or later PR?

Reply:
It could be later, no rush